### PR TITLE
Accept settings in snake case in Docker image

### LINKED ISSUE #74036: docker: allow specification of settings on the commandline without dots
It would be nice to be able to specify settings with environmentvariable not containing dots (`.`) when using the docker container.

Background:
I am using elasticsearch in Microsofts "Azure Container Instances". On this platform, trying to use environment variables containing dots fails with an error:

    Code="InvalidContainerEnvironmentVariable"
    Message="The environment variable name in container 'elasticsearch' of container group 'group' is invalid. A valid environment variable name must start with alphabetic character or '_', followed by a string of alphanumeric characters or '_' (e.g. 'my_name',  or 'MY_NAME',  or 'MyName')."

Concretely I want to pass `discovey.type=single-node`.

I also tried using the environmentvariable `ES_JAVA_OPTS=-Des.discovery.type=single-node` but it seems this particular setting is not picked up from properties.

For Kibana for example the follow statement is done [in the docs](https://www.elastic.co/guide/en/kibana/current/docker.html):

    For compatibility with container orchestration systems, these environment variables are written in all capitals, with underscores as word separators.

### PR DESCRIPTION
**Pull request by pugnascotia**
Closes #74036. Since some orchestration platforms forbid periods in
environment variable names, allow Docker users to pass settings to ES
using an alternative name scheme. For example:

    bootstrap.memory_lock

...becomes:

    ES_BOOTSTRAP_MEMORY__LOCK

The setting name is uppercased, prefixed, all underscores are converted
to double underscores, and all periods are converted to underscores.

### CODE REVIEW

File: distribution/src/bin/elasticsearch-env [hunk 1]
```diff
@@ -122,11 +122,25 @@ if [[ "$ES_DISTRIBUTION_TYPE" == "docker" ]]; then
...
+    # words, e.g. `cluster.name`, or uppercased with underscore separators and
+    # prefixed with `ES_`, e.g. `ES_CLUSTER_NAME`. Underscores in setting names
+    # are escaped by writing them as a double-underscore e.g. "__"
+    if [[ ! -z "$envvar_value" ]]; then
+      if [[ "$envvar_key" =~ ^[a-z0-9_]+\.[a-z0-9_]+ ]]; then
         es_opt="-E${envvar_key}=${envvar_value}"
         es_arg_array+=("${es_opt}")
+      elif [[ "$envvar_key" =~ ^ES(_{1,2}[A-Z]+)+$ ]]; then
+        case "$envvar_key" in
+         ES_JAVA_HOME|ES_PATH_CONF|ES_HOME|ES_JAVA_OPTS|ES_LOG_STYLE|ES_SD_NOTIFY|ES_KEYSTORE_PASSPHRASE_FILE|ES_TMPDIR)
```
Comment (mark-vieira) [inline distribution/src/bin/elasticsearch-env:134]: I'd certainly prefer not to have to manually maintain this list here. Perhaps we could use a unique prefix that doesn't overlap with other environment variables?

Comment (pugnascotia) [reply distribution/src/bin/elasticsearch-env]: I agree in principal, but I've racked my brain and I'm struggling to come up with anything better. Any ideas?

| **Prefix** | **Expands to** | **Drawback** |
|------------|----------------|--------------|
| `ESS_` | ES setting | Sounds like the ESS service |
| `E_` | Elasticsearch | Sounds like an error name |
| `ES_SETTING_` | Self explanatory | So verbose! |
| `ES_CONFIG_` | Self explanatory | So verbose! |
| `ES_CFG_` | Self explanatory | Still verbose, though a little better |
| No prefix! | Same as what Kibana accepts | Can't avoid a collision with other programs |

I liked the consistency of `ES_` for vars that relate to ES - even if that means some maintenance overhead. I don't imagine us adding env vars very often?

For what it's worth I search for var names with this:

```sh
rg -g '!**/*.{java,grooy,gradle}' -g '!**/*.asciidoc' -g '!**/*.bat' '\bES_' | perl -ne 'print "$1\n" if /\b(ES_\w+)/' | sort -u
```

Comment (mark-vieira) [reply distribution/src/bin/elasticsearch-env]: > I don't imagine us adding env vars very often?

Probably not. And if/when we do, and we don't update the list it _should_ explode on us during testing.

I also cannot think of a good non-verbose alternative. Let's push forward with this as-is then.

**Review by mark-vieira — APPROVED** [review]

**Comment by pugnascotia** [comment]
@elasticmachine update branch

**Comment by pugnascotia** [comment]
Backported to `7.x` in f3b25f43c54.

**Pull request merged**